ros2cli/ros2cli/node/daemon.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import platform
+import socket
+import subprocess
+
+from ros2cli.daemon import PORT
+
+import xmlrpc.client
+
+
+def is_daemon_running(args):
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    if platform.system() != 'Windows':
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    addr = ('localhost', PORT)
+    try:
+        s.bind(addr)
+    except socket.error as e:
+        if e.errno in [98, 10048]:
+            return True
+    else:
+        s.close()
+    return False
+
+
+def spawn_daemon(args):
+    subprocess.Popen(
+        ['_ros2_daemon'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
```

Comment:
> Also, what happens if I do?:

```
$ ROS_DOMAIN_ID=42 ros2 topic list
$ ROS_DOMAIN_ID=43 ros2 topic list
```

The daemon and the command line tools currently don't know anything about `ROS_DOMAIN_ID`. So this use case is currently not covered. I can certainly add this though - different ways are possible:

* E.g. whenever a RPC is being made to the daemon the current domain id is being passed. The daemon only answers the request if the domain id matches. Otherwise it should probably shut itself down.

* The daemon chooses a port number which depends on the domain id. That way multiple daemon would happily coexist when a user uses different domain ids from different shells.